File: libs/auth/python/hellopro_auth/oauth.py
"""Framework-free OAuth2 + PKCE primitives (parity with libs/auth/node oauth.ts)."""
from __future__ import annotations

import base64
import hashlib
import secrets
from dataclasses import dataclass
from urllib.parse import urlencode

import httpx
import jwt


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


@dataclass(frozen=True)
class Pkce:
    verifier: str
    challenge: str


def gen_pkce() -> Pkce:
    verifier = _b64url(secrets.token_bytes(32))
    challenge = _b64url(hashlib.sha256(verifier.encode()).digest())
    return Pkce(verifier=verifier, challenge=challenge)


def random_state() -> str:
    return _b64url(secrets.token_bytes(16))


def build_authorize_url(*, public_url: str, client_id: str, redirect_uri: str,
                         challenge: str, state: str) -> str:
    query = urlencode({
        "response_type": "code",
        "client_id": client_id,
        "redirect_uri": redirect_uri,
        "code_challenge": challenge,
        "code_challenge_method": "S256",
        "state": state,
    })
    return f"{public_url}/authorize?{query}"


def exchange_code(*, base_url: str, client_id: str, client_secret: str, code: str,
                   redirect_uri: str, verifier: str, timeout: float = 10.0) -> dict:
    resp = httpx.post(
        f"{base_url}/token",
        data={
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": redirect_uri,
            "code_verifier": verifier,
        },
        auth=(client_id, client_secret),
        headers={"Content-Type": "application/x-www-form-urlencoded"},
        timeout=timeout,
    )
    if resp.status_code != 200:
        raise RuntimeError(f"[account-auth] token exchange failed: {resp.status_code}")
    return resp.json()


@dataclass(frozen=True)
class Identity:
    email: str
    name: str | None = None


def verify_and_extract(access_token: str, jwt_secret: str) -> Identity:
    # aud intentionally not verified (account-service sets aud=client_id).
    payload = jwt.decode(access_token, jwt_secret, algorithms=["HS256"],
                          options={"verify_aud": False})
    email = payload.get("sub") or payload.get("email")
    if not email:
        raise RuntimeError("[account-auth] token missing sub/email claim")
    return Identity(email=email, name=payload.get("name"))

File: libs/auth/node/src/oauth.ts
import { jwtVerify } from "jose"

function b64url(buf: ArrayBuffer | Uint8Array): string {
  const bytes = buf instanceof Uint8Array ? buf : new Uint8Array(buf)
  return Buffer.from(bytes)
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "")
}

export interface Pkce {
  verifier: string
  challenge: string
}

export async function generatePkce(): Promise<Pkce> {
  const verifier = b64url(crypto.getRandomValues(new Uint8Array(32)))
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(verifier))
  return { verifier, challenge: b64url(digest) }
}

export function randomState(): string {
  return b64url(crypto.getRandomValues(new Uint8Array(16)))
}

export function buildAuthorizeUrl(opts: {
  publicUrl: string
  clientId: string
  redirectUri: string
  challenge: string
  state: string
}): string {
  const params = new URLSearchParams({
    response_type: "code",
    client_id: opts.clientId,
    redirect_uri: opts.redirectUri,
    code_challenge: opts.challenge,
    code_challenge_method: "S256",
    state: opts.state,
  })
  return `${opts.publicUrl}/authorize?${params.toString()}`
}

export interface TokenResponse {
  access_token: string
  refresh_token?: string
  token_type?: string
  expires_in?: number
}

export async function exchangeCode(opts: {
  baseUrl: string
  clientId: string
  clientSecret: string
  code: string
  redirectUri: string
  verifier: string
}): Promise<TokenResponse> {
  const basic = Buffer.from(`${opts.clientId}:${opts.clientSecret}`).toString("base64")
  const body = new URLSearchParams({
    grant_type: "authorization_code",
    code: opts.code,
    redirect_uri: opts.redirectUri,
    code_verifier: opts.verifier,
  })
  const r = await fetch(`${opts.baseUrl}/token`, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Authorization: `Basic ${basic}`,
    },
    body: body.toString(),
  })
  if (!r.ok) {
    throw new Error(`[redis-client] token exchange failed: ${r.status}`)
  }
  return (await r.json()) as TokenResponse
}

export interface Identity {
  email: string
  name?: string
}

export async function verifyAndExtract(accessToken: string, jwtSecret: string): Promise<Identity> {
  const key = new TextEncoder().encode(jwtSecret)
  // aud is intentionally NOT verified (account-service sets aud=client_id).
  const { payload } = await jwtVerify(accessToken, key, { algorithms: ["HS256"] })
  const email = (payload.sub as string | undefined) ?? (payload.email as string | undefined)
  if (!email) throw new Error("[redis-client] token missing sub/email claim")
  return { email, name: payload.name as string | undefined }
}
